File: js/ai.js
const DEFAULT_PRIORITY_MATRIX = [
	[15, 14, 13, 12],
	[8,   9, 10, 11],
	[7,   6,  5,  4],
	[0,   1,  2,  3]
];

function AI(weights, priorityMatrix = DEFAULT_PRIORITY_MATRIX){
	this.searchDepth = 1;
	this.weightCount = 6;
	this.weights = [];

	if(weights == undefined){
		for(let i = 0; i < this.weightCount; i++){
			this.weights.push(Math.random());
		}
	}else{
		this.weights = weights;
	}

	this.priorityMatrix = priorityMatrix;
	this.score = 0;
}

AI.prototype.getNextMove = function(gameManager, depth = this.searchDepth){
	let bestHeuristic = undefined;
	let bestHeuristicIndex = -1;

	let available_moves = [];

	for(let i = 0; i <= 3; i++){
		let thisMoves = gameManager.copyToSimplified();
		thisMoves.move(i);

		if(thisMoves.grid.isGridEqual(gameManager.grid))
			continue;

		available_moves.push(i);

		let heuristic = undefined;

		if(depth > 0){
			let availableCells = thisMoves.grid.availableCells();
			for(let ac = 0; ac < availableCells.length; ac++){
				let cell = availableCells[ac];

				thisMoves.grid.insertTile(new Tile(cell, 2));
				let heuristic2Tile = this.getNextMove(thisMoves, depth - 1);

				thisMoves.grid.insertTile(new Tile(cell, 4));
				let heuristic4Tile = this.getNextMove(thisMoves, depth - 1);

				let heuristicVal = heuristic2Tile.heuristicValue * 0.8 + heuristic4Tile.heuristicValue * 0.2;

				if(heuristic == undefined){
					heuristic = heuristicVal;
				}else{
					heuristic = Math.min(heuristicVal, heuristic);
				}

				thisMoves.grid.removeTile(cell);
			}
		}else{
			heuristic = this.getGridScore(thisMoves.grid);
		}

		if(bestHeuristic == undefined || heuristic > bestHeuristic){
			bestHeuristic = heuristic;
			bestHeuristicIndex = i;
		}
	}

	if(bestHeuristicIndex == -1) return {index: available_moves[0], heuristicValue: bestHeuristic};

	return {index: bestHeuristicIndex, heuristicValue: bestHeuristic};
}

AI.prototype.getGridScore = function(grid){
	let sumPriority = 0;
	let sumAdjacentX = 0;
	let sumAdjacentY = 0;
	let numberOpenTiles = 0;
	let maxTile = 0;
	let average = 0;
	let averageCount = 0;

	for(let x = 0; x < grid.size; x++){
		for(let y = 0; y < grid.size; y++){
			let thisCell = grid.cells[x][y];
			let thisCellValue = thisCell ? thisCell.value : 0;

			if(thisCellValue == 0){
				numberOpenTiles++;
			}else if(thisCellValue > maxTile){
				maxTile = thisCellValue;
			}

			if(thisCellValue > 0){
				average += thisCellValue;
				averageCount++;
			}

			sumPriority += Math.pow(this.priorityMatrix[x][y], 2) * thisCellValue;

			let vectors = [
				{x: x - 1, y: y},
				{x: x + 1, y: y},
				{x: x, y: y - 1},
				{x: x, y: y + 1},
			];

			for(let i = 0; i < vectors.length; i++){
				if(grid.withinBounds(vectors[i])){
					let thisLoopCell = grid.cells[vectors[i].x][vectors[i].y];
					let thisLoopCellValue = thisLoopCell ? thisLoopCell.value : 0;

					if(i <= 1){
						sumAdjacentX += Math.abs(thisLoopCellValue - thisCellValue);
					}else{
						sumAdjacentY += Math.abs(thisLoopCellValue - thisCellValue);
					}
				}
			}
		}
	}

	average /= averageCount;

	return (sumPriority * this.weights[0]) +
			 	 (sumAdjacentX * this.weights[1]) +
			   (sumAdjacentY * this.weights[2]) +
				 (maxTile * this.weights[3]) +
				 (numberOpenTiles * this.weights[4]);
			 	 (average * this.weights[5]);
}

File: ai_training_cpp/src/ai.cpp
#include "ai.h"

#define SEARCH_DEPTH 5

AI::AI(){
  this->searchDepth = SEARCH_DEPTH;
  this->score = 0;

  this->priorityWeight = 1.0;
  this->adjacentXWeight = -1.0;
  this->adjacentYWeight = -1.0;
  this->maxTileWeight = 0;
  this->openTilesWeight = 0;
  this->averageWeight = 0;
  this->randomTiles = 0.17;
}

AI::AI(double pri, double adjX, double adjY, double maxT, double openT, double avg, double random){
  this->searchDepth = SEARCH_DEPTH;
  this->score = 0;

  this->priorityWeight = pri;
  this->adjacentXWeight = adjX;
  this->adjacentYWeight = adjY;
  this->maxTileWeight = maxT;
  this->openTilesWeight = openT;
  this->averageWeight = avg;
  this->randomTiles = random;
}

Move AI::getNextMove(Game snapshot){
  return this->getNextMove(snapshot, this->searchDepth);
}

Move AI::getNextMove(Game snapshot, int depth){
  double bestHeuristic = -99999;
  int bestHeuristicIndex = -1;

  for(int i = 0; i <= 3; i++){
    Game thisMoves = snapshot.copy();
    thisMoves.move(i);

    if(thisMoves.grid.isGridEqual(snapshot.grid))
			continue;

    double heuristic = -1;

    for(int j = 0; j < std::floor(this->randomTiles * 6); j++) thisMoves.addRandomTile();

    heuristic = this->getGridScore(thisMoves.grid);

    if(depth > 0){
      Move resultHeuristic = this->getNextMove(thisMoves, depth - 1);
      heuristic += resultHeuristic.heuristicValue;
    }

    if(heuristic > bestHeuristic){
			bestHeuristic = heuristic;
			bestHeuristicIndex = i;
		}
  }

  if(bestHeuristicIndex == -1){
    Move m(rand() % 4, bestHeuristic);
    return m;
  }

  Move m(bestHeuristicIndex, bestHeuristic);
  return m;
}

double AI::getGridScore(Grid grid){
  double sumPriority = 0;
	double sumAdjacentX = 0;
	double sumAdjacentY = 0;
	double numberOpenTiles = 0;
	int maxTile = 0;
	double average = 0;
	double averageCount = 0;

	for(int x = 0; x < 4; x++){
		for(int y = 0; y < 4; y++){
			int thisCellValue = grid.cells[x][y].value;

			if(thisCellValue == 0){
				numberOpenTiles++;
			}else if(thisCellValue > maxTile){
				maxTile = thisCellValue;
			}

			if(thisCellValue > 0){
				average += thisCellValue;
				averageCount++;
			}

			//sumPriority += this.priorityMatrix[x][y] * thisCellValue;
			sumPriority += std::pow(this->priorityMatrix[x][y] * 2, 2) * thisCellValue;

			int vectors[4][2] = {
				{x - 1, y},
				{x + 1, y},
				{x, y - 1},
				{x, y + 1},
			};

			for(int i = 0; i < 4; i++){
				if(grid.withinBounds(Position(vectors[i][0], vectors[i][1]))){
					int thisLoopCellValue = grid.cells[vectors[i][0]][vectors[i][1]].value;

					if(i <= 1){
						sumAdjacentX += std::abs(thisLoopCellValue - thisCellValue);
					}else{
						sumAdjacentY += std::abs(thisLoopCellValue - thisCellValue);
					}
				}
			}
		}
	}

	average /= averageCount;

	return (average * this->averageWeight) + (sumPriority * this->priorityWeight) + (sumAdjacentX * this->adjacentXWeight) + (sumAdjacentY * this->adjacentYWeight) + (maxTile * this->maxTileWeight) + (numberOpenTiles * this->openTilesWeight);
}
